piecrust/src/call_tree.rs

```diff
@@ -0,0 +1,248 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use std::marker::PhantomData;
+use std::mem;
+
+use piecrust_uplink::ContractId;
+
+/// An element of the call tree.
+#[derive(Debug, Clone, Copy)]
+pub struct CallTreeElem {
+    pub contract_id: ContractId,
+    pub limit: u64,
+    pub spent: u64,
+    pub mem_len: usize,
+}
+
+/// The tree of contract calls.
+#[derive(Debug, Default)]
+pub struct CallTree(Option<*mut CallTreeNode>);
+
+impl CallTree {
+    /// Creates a new empty call tree, starting with at the given contract.
+    pub(crate) const fn new() -> Self {
+        Self(None)
+    }
+
+    /// Push an element to the call tree.
+    ///
+    /// This pushes a new child to the current node, and advances to it.
+    pub(crate) fn push(&mut self, elem: CallTreeElem) {
+        match self.0 {
+            None => self.0 = Some(CallTreeNode::new(elem)),
+            Some(inner) => unsafe {
+                let node = CallTreeNode::with_prev(elem, inner);
+                (*inner).children.push(node);
+                self.0 = Some(node)
+            },
+        }
+    }
+
+    /// Moves to the previous node and set the gas spent of the current element,
+    /// returning it.
+    pub(crate) fn move_up(&mut self, spent: u64) -> Option<CallTreeElem> {
+        self.0.map(|inner| unsafe {
+            (*inner).elem.spent = spent;
+            let elem = (*inner).elem;
+
+            let prev = (*inner).prev;
+            if prev.is_none() {
+                free_tree(inner);
+            }
+            self.0 = prev;
+
+            elem
+        })
+    }
+
+    /// Moves to the previous node, clearing the tree under it, and returns the
+    /// current element.
+    pub(crate) fn move_up_prune(&mut self) -> Option<CallTreeElem> {
+        self.0.map(|inner| unsafe {
+            let elem = (*inner).elem;
+
+            let prev = (*inner).prev;
+            if let Some(prev) = prev {
+                (*prev).children.pop();
+            }
+            free_tree(inner);
+            self.0 = prev;
+
+            elem
+        })
+    }
+
+    /// Give the current node the amount spent and recusively update amount
+    /// spent to accurately reflect what each node spent during each call.
+    pub(crate) fn update_spent(&mut self, spent: u64) {
+        if let Some(inner) = self.0 {
+            unsafe {
+                (*inner).elem.spent = spent;
+                update_spent(inner);
+            }
+        }
+    }
+
+    /// Returns the `n`th previous element from the counting from the current
```

Comment:
I guess the first "from the" is a leftover

Reply:
Merging the call stack with the call tree resulted in this. Fixed!